### PR TITLE
`Paywalls`: new `react-native-purchases-ui` for cross-platform paywalls

### PR DESCRIPTION
**Pull request by NachoSoto**
Depends on https://github.com/RevenueCat/purchases-hybrid-common/pull/549

### CODE REVIEW

File: examples/purchaseTesterTypescript/ios/Podfile [hunk 1]
```diff
@@ -57,5 +62,12 @@ target 'PurchaseTester' do
       :mac_catalyst_enabled => false
     )
     __apply_Xcode_12_5_M1_post_install_workaround(installer)
+
+    # See https://github.com/facebook/react-native/issues/37748
```
Comment (NachoSoto) [inline examples/purchaseTesterTypescript/ios/Podfile:66]: This fixes Xcode 15

File: android/src/main/java/com/revenuecat/purchases/react/Paywall.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import com.revenuecat.purchases.ui.revenuecatui.PaywallOptions
+
+internal class Paywall @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+) : AbstractComposeView(context, attrs) {
+    @OptIn(ExperimentalPreviewRevenueCatUIPurchasesAPI::class)
+    @Composable
+    override fun Content() {
+        Paywall(
```
Comment (tonidero) [inline android/src/main/java/com/revenuecat/purchases/react/Paywall.kt:26]: I thought this didn't work? Is this from a previous attempt?

**Review by tonidero — COMMENTED** [review]
Love this! Just holding approval until we solve the issues in PHC.

File: android/src/main/java/com/revenuecat/purchases/react/Paywall.kt (hunk 2 above)
Comment (NachoSoto) [reply android/src/main/java/com/revenuecat/purchases/react/Paywall.kt:26]: Yeah I need to comment it out or disable it. This is how we'll want to support footer view once https://github.com/react-native-community/discussions-and-proposals/issues/446 is resolved.

**Comment by ercanozer** [comment]
@NachoSoto Because of some urgency we decided to use the beta version on production, and it looks working perfectly except one thing. After purchase is complete on paywall, it does not auto, so I applied a patch for now to export a dismissPaywall function like below. So wondering if you guys will add a auto close the paywall support a function too.

```
RCT_EXPORT_METHOD(dismissPaywall) {
    if (@available(iOS 15.0, *)) {
        dispatch_async(dispatch_get_main_queue(), ^{
            UIViewController *rootViewController = [UIApplication sharedApplication].keyWindow.rootViewController;
            if ([rootViewController.presentedViewController isKindOfClass:[RCPaywallViewController class]]) {
                [rootViewController.presentedViewController dismissViewControllerAnimated:YES completion:nil];
            }
        });
    } else {
        // TODO: log
    }
}
```

**Comment by NachoSoto** [comment]
@ercanozer thanks for the report! I just fixed that so it's consistent with the default Android behavior: https://github.com/RevenueCat/purchases-hybrid-common/pull/549/commits/6190aebc2041f190989f6bb82631eb09bd257a0b (cc @tonidero)

I'll release a new beta with that fix :)

**Comment by rikur** [comment]
When will this land? I'm hoping to benefit from the fixes done to ios/android deps.

**Comment by rikur** [comment]
Would it be possible to have a way to use the

```
        purchaseCompleted: PurchaseOrRestoreCompletedHandler? = nil,
        restoreCompleted: PurchaseOrRestoreCompletedHandler? = nil,
        onDismiss: (() -> Void)? = nil
```

callbacks from the JS side? 🙏

**Comment by FelipeLujan** [comment]
I added this to my package.json
`{
  "dependencies": {
    "react-native-purchases-ui": "7.15.0-rc.1"
  }
}`

and started getting this error when 
`expo run:android`


`> Task :app:checkDebugAarMetadata FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:checkDebugAarMetadata'.
> A failure occurred while executing com.android.build.gradle.internal.tasks.CheckAarMetadataWorkAction
   > 2 issues were found when checking AAR metadata:
     
       1.  Dependency 'androidx.emoji2:emoji2-views-helper:1.4.0' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :app is currently compiled against android-33.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.4.2 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).
     
       2.  Dependency 'androidx.emoji2:emoji2:1.4.0' requires libraries and applications that
           depend on it to compile against version 34 or later of the
           Android APIs.
     
           :app is currently compiled against android-33.
     
           Also, the maximum recommended compile SDK version for Android Gradle
           plugin 7.4.2 is 33.
     
           Recommended action: Update this project's version of the Android Gradle
           plugin to one that supports 34, then update this project to use
           compileSdkVerion of at least 34.
     
           Note that updating a library or application's compileSdkVersion (which
           allows newer APIs to be used) can be done separately from updating
           targetSdkVersion (which opts the app in to new runtime behavior) and
           minSdkVersion (which determines which devices the app can be installed
           on).

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/8.0.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 23s
72 actionable tasks: 57 executed, 12 from cache, 3 up-to-date
Error: /Users/felipe/Documents/coding/MyProject/app/android/gradlew exited with non-zero code: 1
Error: /Users/felipe/Documents/coding/MyProject/app/android/gradlew exited with non-zero code: 1
    at ChildProcess.completionListener (/Users/felipe/Documents/coding/MyProject/app/node_modules/@expo/cli/node_modules/@expo/spawn-async/build/spawnAsync.js:52:23)
    at Object.onceWrapper (node:events:632:26)
    at ChildProcess.emit (node:events:517:28)
    at maybeClose (node:internal/child_process:1098:16)
    at ChildProcess._handle.onexit (node:internal/child_process:303:5)
    ...
    at Object.spawnAsync [as default] (/Users/felipe/Documents/coding/MyProject/app/node_modules/@expo/cli/node_modules/@expo/spawn-async/build/spawnAsync.js:17:21)
    at spawnGradleAsync (/Users/felipe/Documents/coding/MyProject/app/node_modules/@expo/cli/build/src/start/platforms/android/gradle.js:72:46)
    at Object.assembleAsync (/Users/felipe/Documents/coding/MyProject/app/node_modules/@expo/cli/build/src/start/platforms/android/gradle.js:52:18)
    at runAndroidAsync (/Users/felipe/Documents/coding/MyProject/app/node_modules/@expo/cli/build/src/run/android/runAndroidAsync.js:36:24)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.`

**Comment by tonidero** [comment]
Hi @FelipeLujan, thanks for reaching out.

Our android paywalls have some dependencies that require updating your `compileSdkVersion` to 34 at least. Are you able to update the `compileSdkVersion` inside your app's `android/build.gradle` file? Let us know if that works.

File: .version [hunk 3]
```diff
@@ -1 +1 @@
-7.4.0-SNAPSHOT
+7.15.0-rc.1
```
Comment (NachoSoto) [inline .version:1]: We need to revert this before merging

File: VERSIONS.md [hunk 4]
```diff
@@ -1,112 +1,117 @@
...
-| 3.0.1   | 3.0.0       | 3.0.3           | 1.0.1                |
-| 3.0.0   | 3.0.0       | 3.0.2           | 1.0.1                |
-| 2.4.1   | 2.6.0       | 2.4.0           | 0.1.4                |
-| 2.4.0   | 2.6.0       | 2.4.0           | 0.1.3                |
-| 2.3.4   | 2.5.0       | 2.3.1           | 0.1.2                |
-| 2.3.3   | 2.5.0       | 2.3.1           | 0.1.2                |
-| 2.3.2   | 2.5.0       | 2.3.0           | 0.1.2                |
-| 2.3.1   | 2.5.0       | 2.3.0           | 0.1.1                |
-| 2.3.0   | 2.4.0       | 2.3.0           | N/A                  |
+| Version       | iOS version | Android version | Common files version  |
```
Comment (NachoSoto) [inline VERSIONS.md:1]: Oops all this file changed somehow. Will revert when rebasing.

File: RNPurchases.podspec [hunk 5]
```diff
@@ -16,14 +16,14 @@ Pod::Spec.new do |spec|
...
   # developers who don't want to use Cocoapods
   spec.exclude_files = [
     "ios/Purchases.framework",
     "ios/PurchasesHybridCommon.framework"
   ]
 
   spec.dependency   "React-Core"
-  spec.dependency   "PurchasesHybridCommon", '7.3.3'
-  spec.swift_version    = '5.0'
+  spec.dependency   "PurchasesHybridCommon", '8.10.0-beta.10'
```
Comment (NachoSoto) [inline RNPurchases.podspec:27]: Gotta revert this.

File: ios/RNPurchases.m [hunk 6]
```diff
@@ -470,7 +472,7 @@ - (NSString *)platformFlavor {
 }
 
 - (NSString *)platformFlavorVersion {
-    return @"7.4.0-SNAPSHOT";
+    return @"7.15.0-rc.1";
```
Comment (NachoSoto) [inline ios/RNPurchases.m:475]: Gotta revert this.

File: package.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "react-native-purchases",
   "title": "React Native Purchases",
-  "version": "7.4.0-SNAPSHOT",
+  "version": "7.15.0-rc.1",
```
Comment (NachoSoto) [inline package.json:4]: Gotta revert this.

File: react-native-purchases-ui/RNPaywalls.podspec [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  spec.homepage     = "https://github.com/RevenueCat/react-native-purchases"
+  spec.license      = package['license']
+  spec.platform     = :ios, "11.0"
+
+  spec.source       = { :git => "https://github.com/RevenueCat/react-native-purchases.git" }
+  spec.source_files = "ios/**/*.{h,m,swift}"
+  spec.pod_target_xcconfig = { 'DEFINES_MODULE' => 'YES' }
+
+  spec.dependency   "React-Core"
+  spec.dependency   "PurchasesHybridCommon", '8.10.0-beta.10'
```
Comment (NachoSoto) [inline react-native-purchases-ui/RNPaywalls.podspec:20]: Gotta change this.

File: react-native-purchases-ui/android/build.gradle [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+def kotlin_version = getExtOrDefault("kotlinVersion")
+
+dependencies {
+  // For < 0.71, this will be from the local maven repo
+  // For > 0.71, this will be replaced by `com.facebook.react:react-android:$version` by react gradle plugin
+  //noinspection GradleDynamicVersion
+  implementation "com.facebook.react:react-native:+"
+  implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+  implementation 'com.revenuecat.purchases:purchases-hybrid-common-ui:8.10.0-beta.10'
```
Comment (NachoSoto) [inline react-native-purchases-ui/android/build.gradle:94]: Gotta change this.

File: react-native-purchases-ui/android/build.gradle [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    }
+  }
+
+  compileSdkVersion getExtOrIntegerDefault("compileSdkVersion")
+
+  defaultConfig {
+    minSdkVersion getExtOrIntegerDefault("minSdkVersion")
+    targetSdkVersion getExtOrIntegerDefault("targetSdkVersion")
+    versionCode 1
+    versionName '7.15.0-rc.1'
```
Comment (NachoSoto) [inline react-native-purchases-ui/android/build.gradle:62]: Gotta change this.

File: react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/RNPaywallsModule.kt [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @ReactMethod
+    fun presentPaywallIfNeeded(requiredEntitlementIdentifier: String?, promise: Promise) {
+        presentPaywall(requiredEntitlementIdentifier, promise)
+    }
+
+    private fun presentPaywall(requiredEntitlementIdentifier: String?, promise: Promise) {
+        val fragment = currentActivityFragment
+            ?: // TODO: log
```
Comment (NachoSoto) [inline react-native-purchases-ui/android/src/main/java/com/revenuecat/purchases/react/ui/RNPaywallsModule.kt:41]: We need to address this.

File: scripts/docs/index.html [hunk 12]
```diff
@@ -2,8 +2,8 @@
 <!DOCTYPE html>
 <html>
 <head>
-	<meta http-equiv="refresh" content="0; url=https://revenuecat.github.io/react-native-purchases-docs/7.4.0-SNAPSHOT/" />
+	<meta http-equiv="refresh" content="0; url=https://revenuecat.github.io/react-native-purchases-docs/7.15.0-rc.1/" />
```
Comment (NachoSoto) [inline scripts/docs/index.html:5]: Gotta change this

File: react-native-purchases-ui/package.json [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        "typescript",
+        {
+          "project": "tsconfig.build.json"
+        }
+      ]
+    ]
+  },
+  "dependencies": {
+    "@revenuecat/purchases-typescript-internal": "8.10.0-beta.10",
+    "react-native-purchases": "7.15.0-rc.1"
```
Comment (NachoSoto) [inline react-native-purchases-ui/package.json:113]: Gotta change these

File: react-native-purchases-ui/package.json [hunk 14]
```diff
@@ -0,0 +1,115 @@
+{
+  "name": "react-native-purchases-ui",
+  "title": "React Native Purchases UI",
+  "version": "7.15.0-rc.1",
```
Comment (NachoSoto) [inline react-native-purchases-ui/package.json:4]: Change this

**Pull request closed without merging**